### DISCUSSION TITLE
Multiple logs within same api

### DISCUSSION THREAD

**Discussion by TahaShahabuddin** [thepirat000/Audit.NET#600] [Show and tell]
Hi @thepirat000 

while using audit web api, need to add multiple logs in same api method, 

var auditScope = this.GetCurrentAuditScope();_

creating above scope and adding a comment in that api,

so similarly need to add two different logs 

tried to add multiple logs but its like one instance is only added in the logs the last one in the flow and the previous one are ignored

regards.

**Comment by thepirat000** [comment]
Please share the code where you're trying to do multiple logs.

Note that each AuditScope instance is related to only one audit event.

**Reply by TahaShahabuddin** [reply]
public void Addlogs(AuditLogsInput Input)
        {
            var auditScope = this.GetCurrentAuditScope();
            auditScope.Comment(Input.comment);
            auditScope.SetCustomField(AuditCustomField.source_table_id_uuid, Input.source_table_id_uuid);
            auditScope.SetCustomField(AuditCustomField.module_table_link_id, Input.module_table_link_id);
            auditScope.SetCustomField(AuditCustomField.ip_address, Input.ip_address);
            auditScope.SetCustomField(AuditCustomField.browser_name, Input.browser_name);
        }

       [HttpPost("api")]
        public async void ApiFunc(ClInput )
        {
         Addlogs(new AuditLogsInput { comment = "comment 1 for first event", source_table_id_uuid = Input.merchant_account_id,    module_table_link_id = (int)enumModuleLink.organizations, ip_address = res.IpAddress, browser_name = Request.Cookies[AuditCookies.browser_name]);

          Addlogs(new AuditLogsInput { comment = "comment 2 for different event within same api" , source_table_id_uuid = Input.merchant_account_id, module_table_link_id = (int)enumModuleLink.organizations, ip_address = res.IpAddress, browser_name = Request.Cookies[AuditCookies.browser_name]);
}


@thepirat000  in above code created a class addlogs which has auditscope instance, when we are calling this one time in api it save the log,
but in any api if we are calling it twice it will only save one log not the other, in above case the second one will be saved, 
now here i need a solution so that i can save multiple logs with in same api

regards

**Comment by thepirat000** [comment]
An AuditScope instance is for a single AuditEvent, so you cannot use the same scope to store different audit events. 

However, you could have a custom list field on your API audit event and store your multiple table events in the same AuditEvent like this.
```c#
private void AddLogs(AuditLogsInput input)
{
    var scope = this.GetCurrentAuditScope();
    if (!scope.Event.CustomFields.ContainsKey("InputEvents"))
    {
        scope.Event.CustomFields["InputEvents"] = new List<AuditLogsInput>();
    }
    // Include the table event input in the audit event custom list field
    ((List<AuditLogsInput>)scope.Event.CustomFields["InputEvents"]).Add(input);
}
```

So you'll end up having events like this:

```javascript
{
  "Action": {
    "TraceId": "40000086-0003-ff00-b63f-84710c7967bb",
    "ControllerName": "Values",
    "ActionName": "Get",
    ...
  },
  "EventType": "MVC",
  "Environment": {
	  ...
  },
  "InputEvents": [
    {
      "comment": "test1",
      "source_table_id_uuid": "table1",
      "module_table_link_id": 1,
      "ip_address": "1.2.3.4",
      "browser_name": "test1"
    },
    {
      "comment": "test2",
      "source_table_id_uuid": "table2",
      "module_table_link_id": 2,
      "ip_address": "4.3.2.1",
      "browser_name": "test2"
    }
  ]
}
```

Another possibility is to create a new audit scope for each of your table events and correlate the table events with the API events with some correlation value, for example:

```c#
private void AddLogs(AuditLogsInput input)
{
    var scope = this.GetCurrentAuditScope();
    var apiAction = scope.Event.GetWebApiAuditAction();
    // Create and save the table event right away
    AuditScope.Log("TableEvent", new 
    { 
        TraceId = apiAction.TraceId, 
        InputEvent = input 
    });
}
```

**Reply by TahaShahabuddin** [reply]
@thepirat000  well sometime its not correlated events or table events
both the events within a method are different so need to add two different logs in the history table
can we handle such situation in any way,

and why we restricted it to single log.

regards

**Reply by thepirat000** [reply]
It's not a restriction, but a design decision.
There are no restrictions on what the AuditEvent can contain, you can even inherit and extend AuditEvent at will.

Also, the component in charge of storing the audit events is the data provider, and you can completely override the data provider to insert multiple times for a single scope, or whatever is needed.

In the end, it really depends on how and where are you storing the audit events, you could share your Data Provider configuration.

I've shown you two ways you could handle multiple logs; the first one adding your extra logs to the Web API audit event, and the second creating new audit event for each extra log.
